libdrgn/kallsyms.c

```diff
@@ -0,0 +1,625 @@
+ * table as described above. The first decoded byte is a character representing
+ * what type of symbol this is (e.g. text, data structure, etc).
+ */
+struct kallsyms_reader {
+	uint32_t num_syms;
+	uint8_t *names;
+	char *token_table;
+	uint16_t *token_index;
+	bool long_names;
+};
+
+/*
+ * Kallsyms doesn't include symbol length.  We determine symbol length by the
+ * start of the subsequent symbol.  Unfortunately, there can be large gaps in
+ * the symbol table, for instance on x86_64 the Linux kernel has percpu symbols
+ * near the beginning of the address space, and a large gap before normal kernel
+ * symbols.  The result of this is that we can create symbols with incredibly
+ * large sizes, and then drgn's symbolization will print addresses using that
+ * symbol and a very large offset, which is absolutely meaningless.
+ *
+ * To avoid this, we set a cap on the length of a symbol. Unfortunately, this is
+ * a heuristic. It's entirely possible to have very large data symbols. This
+ * value is chosen somewhat arbitrarily, but seems to produce decent results.
+ */
+#define MAX_SYMBOL_LENGTH 0x10000
+
+/*
+ * Since 73bbb94466fd3 ("kallsyms: support "big" kernel symbols"), the
+ * "kallsyms_names" array may use the most significant bit to indicate that the
+ * initial element for each symbol (normally representing the number of tokens
+ * in the symbol) requires two bytes.
+ *
+ * Unfortunately, that means that values 128-255 are now ambiguous: on older
+ * kernels, they should be interpreted literally, but on newer kernels, they
+ * require treating as a two byte sequence. Since the commit included no changes
+ * to the symbol names or vmcoreinfo, there's no way to detect it except via
+ * heuristics.
+ *
+ * The commit in question is a new feature and not likely to be backported to
+ * stable, so our heuristic is that it was first included in kernel 6.1.
+ * However, we first check the environment variable DRGN_KALLSYMS_LONG: if it
+ * exists, then we use its first character to determine our behavior: 1, y, Y
+ * all indicate that we should use long names. 0, n, N all indicate that we
+ * should not.
+ */
+static bool guess_long_names(struct drgn_program *prog)
+{
+	const char *env = getenv("DRGN_KALLSYMS_LONG");
+	const char *osrelease;
+	int i;
+	int major = 0, minor = 0;
+
+	if (env) {
+		if (*env == '1' || *env == 'y' || *env == 'Y')
+			return true;
+		else if (*env == '0' || *env == 'n' || *env == 'N')
+			return false;
+	}
+
+	osrelease = prog->vmcoreinfo.osrelease;
+	for (i = 0; i < sizeof(prog->vmcoreinfo.osrelease) && osrelease[i]; i++) {
+		char c = osrelease[i];
+		if (c < '0' || c > '9')
+			break;
+		major *= 10;
+		major += osrelease[i] - '0';
+	}
+	for (i = i + 1; i < sizeof(prog->vmcoreinfo.osrelease) && osrelease[i] && osrelease[i] != '.'; i++) {
+		char c = osrelease[i];
+		if (c < '0' || c > '9')
+			break;
+		minor *= 10;
+		minor += osrelease[i] - '0';
+	}
```

Comment:
The version number parsing in `orc_version_from_osrelease()` is a little more concise, please copy that. (I debated turning it into a helper function for 3e7244d25fd12a2d497f40fdde2e8e9157d4e1a1, but I decided that I'd prefer to keep it painful to do version checks to discourage it whenever possible.)